### ISSUE TITLE
Incorporation of Microsoft Fluent Design

### ISSUE THREAD

**Issue by Samuel12321** [ShankarBUS/ModernFlyouts#66]
Review from store:
> "Hello! Another fine addition in my Microsoft Store apps collection! I wish the music flyout would have been more "fluent-ish", featuring Microsoft Fluent Design. For now, it feels like more similar to Google Material Design for Chrome/Android."

has also been 2 similar reviews

**Comment by ShankarBUS** [comment]
**"Reveal effect"** - not a big deal (could be easily done using `RadialGradientBrush`)
**"Acrylic effect"** - exploring the possibilities

I don't really get what the mean by **"Fluent Design"** or **"Material Design"**. Our app not even a bit looks like **"Material Design"**

![Material Design In XAML](http://materialdesigninxaml.net/images/demo-buttons.png
)
_Material Design In XAML_

![Fluent Design - ModernWpf](https://github.com/Kinnara/ModernWpf/raw/master/docs/images/Controls.Light.png)
_Fluent Design - ModernWpf_

Ours look somewhat similar to MS Edge. Those two effects are just a small part of **"Fluent Design"**

The new **Xbox store app** has acrylic but it looks completely diverged from **"Fluent Design"**

![The new Xbox Store](https://www.kotaku.com.au/content/uploads/sites/3/2020/08/04/new-microsoft-store-xbox-1.jpg)
_The new Xbox Store_

Even MS doesn't know what **"Fluent Design"** means

I tried to align to this as much I can 

![Windows 10X SMTC panel](https://user-images.githubusercontent.com/63704247/88851826-fdee9a00-d20a-11ea-834f-a9450a1ce79e.png)
_Windows 10X SMTC panel_

![ModernFlyouts](https://github.com/ShankarBUS/ModernFlyouts/raw/master/docs/images/Audio_Session_Music_NoTop.png)

So all that matters is the fancy **"Acrylic"** effect 😂😟

**Comment by ShankarBUS** [comment]
Hey Guys!

I want some design ideas/suggestions for future releases. I have uploaded the design to Figma - https://www.figma.com/file/Gq5AKs5k8lOl0rmDjQeXT2/ModernFlyouts

This is the **Current Design :**

![ModernFlyouts_Mockup_Current](https://user-images.githubusercontent.com/63704247/94680853-0dd84680-0340-11eb-98a2-f0b04628edf7.png)

Some one in the MS Store review suggested an option to hide the thumbnail on the background. This is how it would look like:

![ModernFlyouts_Mockup_NoThumnail](https://user-images.githubusercontent.com/63704247/94681395-ff3e5f00-0340-11eb-9eaa-9436ec829863.png)

Hey @Samuel12321, @Poopooracoocoo, @Cyberdroid1

Can you guys make mockups/give some suggestions of the design you'd like

What should we do regarding #6, #49, #54, #60?

**Comment by ShankarBUS** [comment]
This is how our flyout will look like if it had the fancy **"Acrylic"** effect

![image](https://user-images.githubusercontent.com/63704247/94827233-d04de900-0425-11eb-8695-89b8fdc24f88.png)

**Comment by YanovskyOr** [comment]
> This is how our flyout will look like if it had the fancy **"Acrylic"** effect
> 
> ![image](https://user-images.githubusercontent.com/63704247/94827233-d04de900-0425-11eb-8695-89b8fdc24f88.png)

PLEASE MAKE THIS HAPPEN 🙏

**Comment by Samuel12321** [comment]
> This is how our flyout will look like if it had the fancy **"Acrylic"** effect
> 
> ![image](https://user-images.githubusercontent.com/63704247/94827233-d04de900-0425-11eb-8695-89b8fdc24f88.png)

Looks Amazing😊
I still think it would look better and follow Win10X design better if it had the small space between the media and  volume flyouts.

**Comment by ShankarBUS** [comment]
@Samuel12321,

> Looks Amazing😊

It's just a mock-up though 😁

> I still think it would look better and follow Win10X design better if it had the small space between the media and  volume flyouts.

Like how it previously was?

I was having some troubles with showing shadows. They were casting on the one above. So I had to change the design to host all the contents on a single shadow chrome.

We can reconsider this gap proposal after we implement acrylic successfully.

I kinda fount a way to have acrylic in WPF (with custom blur radius, custom effects, custom shape not rectangle only and finally it doesn't lag while dragging)

But I don't know how we're are going to compose the WPF content over the acrylic. It would take some though. Maybe post v0.9?

**Comment by Poopooracoocoo** [comment]
>and finally it doesn't lag while dragging

isn't that a bug that Windows' introduced though?

**Comment by ShankarBUS** [comment]
Yes, it will happen if we used `SetWindowCompositionAttribute` and `ACCENT_ENABLE_ACRYLICBLURBEHIND`.

I used the **"Visual Layer"** (aka Windows.UI.Composition APIs) and make a impl of acrylic brush myself.

Only hard thing is that, we have to build the acrylic brush from the base like how WinUI does (Backdrop Brush -> Gaussian Blur -> Noise -> Tint -> Luminosity)

And the window should be border less and transparent with no content in it. The Composition APIs will render atop of the window (over the WPF content) but they don't block input. I'm confused about the implementation process.

![Acrylic](https://i.imgur.com/6CEA7oq.jpg)

I don't know if hosting a child window will tackle this. Will report to you once done

**Comment by Poopooracoocoo** [comment]
I think [TranslucentTB](https://github.com/TranslucentTB/TranslucentTB) has some work in that area. Might be worth checking that out.

**Comment by ShankarBUS** [comment]
Seems like they use `SetWindowCompositionAttribute`. As the TaskBar is static and won't be moved, there won't be any problems. But in our case we have to animate the flyout (no fading or sliding :( ). So unfortunately that method won't work for us.

We have to use the method I suggested.

**Comment by ShankarBUS** [comment]
Wew!

Made some progress on the acrylic effect. Now I can render WPF content on the blurred window.

![image](https://user-images.githubusercontent.com/63704247/95679144-6cc67700-0bee-11eb-8514-774d59b6a526.png)

**Comment by ShankarBUS** [comment]
**Update:** Got the WPF content to refresh itself. But the **CPU** & **GPU** usage skyrockets.

![image](https://user-images.githubusercontent.com/63704247/95680174-33ddd080-0bf5-11eb-974a-5ad0e53956b5.png)

If there are some experts out there, please help me to reduce the **CPU** usage and improve the performance.

If we make this effect efficient, then we can easily add this to ModernFlyouts easily with less amount of work.

**Comment by Cyberdroid1** [comment]
Do the values increase when it the window is created, or for the time the window is present?

**Comment by ShankarBUS** [comment]
The usage will be high until the window is closed.

We are re-rendering the WPF content (which is not visible due to transparency) into a byte array and rendering that onto the composition layer (aka the layer which hosts the acrylic).

So, CPU usage in this process is unavoidable.

It would be nice if we can directly pass the WPF buffer into the Composition layer instead of 2 intermediate steps which includes CPU.

**Comment by Cyberdroid1** [comment]
@ShankarBUS The tray icon popup of Dopamine music player produces an acrylic window blur for its window. Have a look at that too.

**Comment by WindowsFanBoi** [comment]
Hey @ShankarBUS,

How about using two windows, one (say AcrylicWindow) would be using the Composition APIs to create the acrylic effect and the second is what we already have (i.e. FlyoutWindow a transparent alpha blended WPF window)

You can then make the AcrylicWindow as the owner of the FlyoutWindow so that they have the same Z-Order and will look a single window if their bounds are synced.

The FlyoutWindow won't consume that much CPU/GPU in idle state and since the AcrylicWindow is using the Composition APIs, it will result in a performance improvement.

But you'd have some problems using Win2D.uwp with .NET 5 (as Win2D will provide *.winmd metadata files not *.dll assemblies and also the WinRT interop has been removed from .NET 5, you won't be able to use them without making your own projection lib using CsWinRT)

So 🤷‍♂️

**Comment by ShankarBUS** [comment]
@WindowsFanBoi

Same pinch!

You're about two hours late 😁, I got it working without any performance degradation. Thanks for the suggestion though.

I didn't there were such problems with .NET 5, I was working on a experimental project which uses .NET Framework 4.8. So, I didn't experience this issue. I will see what we could do about this.

**Comment by Samuel12321** [comment]
don't know if this could be of any help: https://github.com/sourcechord/FluentWPF/tree/master

**Comment by ShankarBUS** [comment]
Bruh! https://github.com/ShankarBUS/ModernFlyouts/issues/44#issuecomment-695158147

**Comment by Samuel12321** [comment]
whops, i forgot all about that one

**Comment by ShankarBUS** [comment]
@Samuel12321,

You said you know C++, right?

Win2D is not supported in .NET 5 (due to various reasons). So we have to make our own implementation with just the GaussianBlurEffect in C# and use it instead of Win2D.

Do you know anything about Direct2D? And how it use in from C#?

Anyone in this thread?

Currently I can use the Composition APIs to render stuff but can't do apply any effects due to the lack of Win2D.

So, just the screen pixels are rendered without the blur.

One way to tackle this would be to fallback to .NET Core 3.1 which I don't prefer.

Any help would be appreciated 🙏.

**Comment by Samuel12321** [comment]
no sorry, as i said, i know python and a specific variant of c (used for controlling mechanical devices), this is my first time with C++, C#, .NET and app development in general.

I will see what i can learn about this though. 
Community help would be very useful.

**Comment by Samuel12321** [comment]
Just renamed thread so it's specific to the acrylic (which is all that seems to be discussed here anyway), will create a separate thread for the Reveal effect.

**Comment by ShankarBUS** [comment]
> Win2D is not supported in .NET 5 (due to various reasons). So we have to make our own implementation with just the GaussianBlurEffect in C# and use it instead of Win2D.
> 
> Do you know anything about Direct2D? And how it use in from C#?
> 
> Anyone in this thread?
> 
> Currently I can use the Composition APIs to render stuff but can't do apply any effects due to the lack of Win2D.
> 
> So, just the screen pixels are rendered without the blur.
> 
> One way to tackle this would be to fallback to .NET Core 3.1 which I don't prefer.
> 
> Any help would be appreciated 🙏.

Nah, no need for any of that. I forked Win2D and modified to be compatible with .NET 5.

@WindowsFanBoi,

> But you'd have some problems using Win2D.uwp with .NET 5 (as Win2D will provide *.winmd metadata files not managed .NET *.dll assemblies and the WinRT interop being removed in .NET 5, you won't be able to use them without making your own projection lib using CsWinRT)

That's what I did.

And the result turned out to be great!

![Acrylic](https://user-images.githubusercontent.com/63704247/100478090-3007fe00-3110-11eb-8b6e-9cc95ec8ce34.gif)

For those who are willing to test this out by themselves,

[WPF Acrylic effect demo.zip](https://github.com/ShankarBUS/ModernFlyouts/files/5609278/WPF.Acrylic.effect.demo.zip)

1. System Requirements:
    - Windows 10 1903 or later
    - x64 OS
    - .NET 5 runtime preinstalled
2. Download and extract this zip file (safe to download and extract, I assure you)
3. Run the `Temp.exe` and voila!

This is just a prototype. I will make some improvement later this week.
You can drag the window by holding mouse left and moving the cursor (just like our flyouts).

For some weird reasons the `SetWindowCompositionAttribute` API (used for having acrylic effect in Win32 apps for ex: used in FluentWPF, AvaloniaUI, EarTrumpet, TranslucentTB, TaskbarX and so on) works fine in Windows 10 20H2 without the so called "1903 drag issue".

The only benefits my approach have are custom blur radius, custom effects (not just blur), custom acrylic visual shape (we could even create a round acrylic window with my approach).
Otherwise, the `SetWindowCompositionAttribute` API is enough.

Feedbacks are welcome!

**Comment by Samuel12321** [comment]
WELLDONE @ShankarBUS 
looks great.

**Comment by Cyberdroid1** [comment]
> For some weird reasons the `SetWindowCompositionAttribute` API (used for having acrylic effect in Win32 apps for ex: used in FluentWPF, AvaloniaUI, EarTrumpet, TranslucentTB, TaskbarX and so on) works fine in Windows 10 20H2 without the so called "1903 drag issue".

Maybe MS finally fixed it...?

Congrats though.

**Comment by TheXamlGuy** [comment]
Following this Twitter conversation (https://twitter.com/TheXamlGuy/status/1341098062245183488) I found another way to do this by using WindowXamlHost to host a UWP usercontrol inside a WPF app and then using a Flyout with it ShouldConstrainToRootBounds property set to false, this will literally allow the Flyout to go outside the root boundary of the window hosting the control. A bonus is is that you are given the animations and acrylic by default. 

I have a quick dirty sample of the above working, I'll get it posted once I have tidied it up some more.

https://user-images.githubusercontent.com/14894973/102831102-d8914100-43e2-11eb-8b58-454f75f33c61.mp4

**Comment by TheXamlGuy** [comment]
Obviously I should point out the only con about the above is that you'd lose the ability to drag the Flyout around, but I am sure with some more work it could be made possible to drag the Flyout around too?

**Comment by ShankarBUS** [comment]
@TheXamlGuy,

Thank you so much mate!!!!!

I will investigate this further.

But **"Window Activation"** is another thing to worry about.

The current flyout is designed to be non activatable and won't activate on mouse activation or even while showing.

I'll sure explore the possibilities.

And how did you find that he was using a **"Flyout"** inside the xaml island?

**Comment by TheXamlGuy** [comment]
The animation, it was far too smooth to have been achieved with a WPF window alone especially with it being a larger window. The guys who work on EarTrumpet have been trying to replicate the look and the smooth animation found the Windows 10 notification flyout for a very long time without any luck. 
I thought at first he was using the TeachingTip control from WinUI  as Microsoft themselves uses the TeachingTip control on the Desktop in areas where they are trying to promote their new Edge browser when someone had Chrome installed. That's when I discovered ShouldConstrainToRootBounds in the FlyoutBase.

**Comment by Tropix126** [comment]
Any updates on this?

**Comment by Samuel12321** [comment]
Hi @Tropix126, not that I'm aware of. This is a bit out side of my area of expertise and i don't think @ShankarBUS was planning on doing anything more on this one.
As always if anyone else wanted to help add this it would be greatly appreciated.

**Comment by Cyberdroid1** [comment]
Using Powertoys Keyboard manager is better, it sometimes produces Modernflyouts instead of the default, but not every time. Try giving it a try

**Comment by Samuel12321** [comment]
Hi Everyone as we are moving to XAML Islands this should now be possible. 
Tentatively adding this to the 0.10 Release

**Comment by SelastinGeorge** [comment]
> > Win2D is not supported in .NET 5 (due to various reasons). So we have to make our own implementation with just the GaussianBlurEffect in C# and use it instead of Win2D.
> > Do you know anything about Direct2D? And how it use in from C#?
> > Anyone in this thread?
> > Currently I can use the Composition APIs to render stuff but can't do apply any effects due to the lack of Win2D.
> > So, just the screen pixels are rendered without the blur.
> > One way to tackle this would be to fallback to .NET Core 3.1 which I don't prefer.
> > Any help would be appreciated 🙏.
> 
> Nah, no need for any of that. I forked Win2D and modified to be compatible with .NET 5.
> 
> @WindowsFanBoi,
> 
> > But you'd have some problems using Win2D.uwp with .NET 5 (as Win2D will provide *.winmd metadata files not managed .NET *.dll assemblies and the WinRT interop being removed in .NET 5, you won't be able to use them without making your own projection lib using CsWinRT)
> 
> That's what I did.
> 
> And the result turned out to be great!
> 
> ![Acrylic](https://user-images.githubusercontent.com/63704247/100478090-3007fe00-3110-11eb-8b6e-9cc95ec8ce34.gif)
> 
> For those who are willing to test this out by themselves,
> 
> [WPF Acrylic effect demo.zip](https://github.com/ShankarBUS/ModernFlyouts/files/5609278/WPF.Acrylic.effect.demo.zip)
> 
> 1. System Requirements:
>    
>    * Windows 10 1903 or later
>    * x64 OS
>    * .NET 5 runtime preinstalled
> 2. Download and extract this zip file (safe to download and extract, I assure you)
> 3. Run the `Temp.exe` and voila!
> 
> This is just a prototype. I will make some improvement later this week.
> You can drag the window by holding mouse left and moving the cursor (just like our flyouts).
> 
> For some weird reasons the `SetWindowCompositionAttribute` API (used for having acrylic effect in Win32 apps for ex: used in FluentWPF, AvaloniaUI, EarTrumpet, TranslucentTB, TaskbarX and so on) works fine in Windows 10 20H2 without the so called "1903 drag issue".
> 
> The only benefits my approach have are custom blur radius, custom effects (not just blur), custom acrylic visual shape (we could even create a round acrylic window with my approach).
> Otherwise, the `SetWindowCompositionAttribute` API is enough.
> 
> Feedbacks are welcome!

@ShankarBUS Can you please share the source code of demo you mentioned here https://github.com/ModernFlyouts-Community/ModernFlyouts/issues/66#issuecomment-734953522
